### PR TITLE
Keyscanning module initial merge

### PR DESCRIPTION
**Pull request by hiszd**
Keyscanning module now functional, with false positives for Pa14 on Gemini.
After porting the embedded-hal 1.00-alpha into atsam4-hal we can resolve this issue.

### CODE REVIEW

File: kiibohd-keyscanning/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+version = "0.0.1"
+authors = ["Zion Koyl <hiszd1@gmail.com>"]
+license = "MIT OR Apache-2.0"
+edition = "2018"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+rtcc            = { version = "^0.1.0" }
+cortex-m        = "0.6"
```
Comment (haata) [inline kiibohd-keyscanning/Cargo.toml:12]: I don't use cortex-m, cortex-m-rt or cortex-m-rtic used?

File: kiibohd-keyscanning/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+cortex-m-rt     = { version = "0.6.6", features = ["device"] }
+cortex-m-rtic   = "0.6.0-alpha.2"
+generic-array   = "0.14.4"
+embedded-hal    = "0.2.5"
+embedded-time   = "0.10.1"
+keyberon        = "0.1.1"
+heapless        = "0.5"
```
Comment (haata) [inline kiibohd-keyscanning/Cargo.toml:19]: I don't see heapless used.

File: kiibohd-keyscanning/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+[dependencies.atsam4-hal]
```
Comment (haata) [inline kiibohd-keyscanning/Cargo.toml:21]: Shouldn't be necessary to include atsam4-hal.

File: kiibohd-keyscanning/README.md [hunk 4]
```diff
@@ -0,0 +1,12 @@
+# hid-io-kiibohd
```
Comment (haata) [inline kiibohd-keyscanning/README.md:1]: Not the correct README.md?

File: kiibohd-keyscanning/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+#![no_std]
+#[allow(unused_imports)]
+#[allow(unused_attributes)]
+/// This crate is to handle scanning and strobing of the key matrix.
+/// It also handles the debouncing of key input to ensure acurate keypresses are being read.
+/// InputPin's, and OutputPin's are passed in through the "rows" and "cols" parameters in the Scan::new() function.
+/// The maximum number of rows is 7, and the maximum number of columns is 20. This number may need adjusted through testing.
+pub mod state;
+pub use self::state::{KeyState, State, StateReturn};
+// use atsam4_hal::{gpio::*, prelude::*, InputPin, OutputPin};
```
Comment (haata) [inline kiibohd-keyscanning/src/lib.rs:17]: Remove

File: kiibohd-keyscanning/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+    where
+        &'a mut C: IntoIterator<Item = &'a mut dyn OutputPin<Error = E>>,
+    {
+        for c in self.cols.into_iter() {
+            c.set_low().ok().unwrap();
+        }
+        Ok(())
+    }
+
+    pub fn get<'a, E: 'a>(&'a mut self, callback: fn(StateReturn, usize, bool)) -> Result<(), E>
```
Comment (haata) [inline kiibohd-keyscanning/src/lib.rs:61]: Add a doc comment here explaining what you are doing (basically your scanning methodology).

File: kiibohd-keyscanning/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+        KeyState::poll_update(&mut self.keys[r][c], high)
+    }
+
+    // Get the individual state of a specific key
+    pub fn get_state(&self, r: usize, c: usize) -> State {
+        KeyState::get_state(&self.keys[r][c])
+    }
+
+    pub fn get_scancode(&self, r: usize, c: usize) -> usize {
+        c + (19 * r)
```
Comment (haata) [inline kiibohd-keyscanning/src/lib.rs:118]: Why 19?

File: kiibohd-keyscanning/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+            c.set_low().ok().unwrap();
+        }
+
+        Ok(())
+    }
+}
+
+/// The matrix to keep all the key states and handle state updating
+pub struct StateMatrix {
+    keys: [[KeyState; 7]; 20],
```
Comment (haata) [inline kiibohd-keyscanning/src/lib.rs:90]: Why 7 and 20?

File: kiibohd-keyscanning/src/matrix.rs [hunk 9]
```diff
@@ -0,0 +1,151 @@
+// Copyright 2021 Zion Koyl
```
Comment (haata) [inline kiibohd-keyscanning/src/matrix.rs:1]: Is this a duplicate file to lib.rs?

**Review by haata — CHANGES_REQUESTED** [review]

File: kiibohd-keyscanning/src/lib.rs (hunk 5 above)
Comment (hiszd) [reply kiibohd-keyscanning/src/lib.rs:17]: If I remove lines 15, and 16, then it won't compile.
I made the state function a sub-module and all of the matrix stuff in the main module.
Are you suggesting that I move all of the code to the main module?

File: kiibohd-keyscanning/src/lib.rs (hunk 7 above)
Comment (hiszd) [reply kiibohd-keyscanning/src/lib.rs:118]: For this keyboard the scan codes start at the top left and increment up from left to right. There are 19 columns, so for each row the scan code would be column + (19 * row).
For the final implementation I would either have this method in the bin.rs for each keyboard, or have a "feature" for each keyboard(which isn't the best option)

File: kiibohd-keyscanning/src/lib.rs (hunk 8 above)
Comment (hiszd) [reply kiibohd-keyscanning/src/lib.rs:90]: These are the preliminary maximum numbers for the rows and columns.
These can be changed to any number we deem as the true maximum.
The only downside would be memory usage.

File: kiibohd-keyscanning/src/matrix.rs (hunk 9 above)
Comment (hiszd) [reply kiibohd-keyscanning/src/matrix.rs:1]: It is. This should have been removed. Removing in next commit.

File: kiibohd-keyscanning/Cargo.toml (hunk 2 above)
Comment (hiszd) [reply kiibohd-keyscanning/Cargo.toml:15]: I am using heapless in lib.rs

**Pull request merged**